Issues #198 ribbon hystrix

marcingrzejszczak opened the pull request:
Fixes the Sleuth Feign support issue related to not closing spans that were interrupted by an exception

cc @dsyer , @spencergibb, @adriancole

spencergibb commented:
I worry about the Feign integration (I think it predates this PR). The moment someone specifies a `@FeignClient(configuration = MyConfig.class)` and specifies a `Decoder` or `Builder` (ie non-hystrix), all of this is broken. Maybe that is something we can live with or document.

marcingrzejszczak commented:
Yeah the easiest would be to make the impls public (the methods not necessarily) and write in the docs that you have to wrap your implementations in our decorators... I was thinking also about a post processor that would wrap all of these

marcingrzejszczak commented:
How about now? I've added some docs, some classes are public so one can wrap his own decoder / retryer in a sleuth solution. Also I've added a bean post processor that wraps those things in our custom impls.

marcingrzejszczak commented:
Any comments? @dsyer @spencergibb @adriancole ? If not then I'd merge this

adriancole commented on spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/feign/TraceFeignRetryer.java:
```diff
@@ -0,0 +1,66 @@
+	public void continueOrPropagate(RetryableException e) {
+		try {
+			this.feignRequestContext.putSpan(this.tracer.getCurrentSpan(), true);
+			this.tracer.getCurrentSpan().logEvent("retry");
```
"feign.retry"

adriancole commented on spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/feign/TraceFeignRequestInterceptor.java:
```diff
@@ -0,0 +1,122 @@
+		publish(new ClientSentEvent(this, span));
+	}
+
+	protected String getSpanName(RequestTemplate template) {
```
I'd always envisioned `MethodMetadata.configKey()` being used for operation name lookups, but that's not available at this call site.. Logger is another way to hack this stuff in

adriancole commented:
I see what you have going on, but it seems like a lot of work. The additional scope of the time spent decoding doesn't seem enough value to justify this much code.

Since tracing is often very similar to logging, I'd recommend just making a Logger decorator that handles these facets. There are a couple package-protected methods that we could open up: logRetry, logIOException.

ex
```java
class TracingLogger extends Logger {
  private final Logger delegate;

// override all methods, tracing, then delegating
--snip--
```

adriancole commented on spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/feign/TraceFeignDecoder.java:
```diff
@@ -0,0 +1,82 @@
+ *
+ * @since 1.0.0
+ */
+public class TraceFeignDecoder implements Decoder, ApplicationEventPublisherAware {
```
since not all responses are decoded, you might need to play with Response.Body itself. Ex closing the span on body.close unless there's no body.

adriancole commented:
PS totally fine to have some of these things refactored later, particularly if types are hidden while they shape up

marcingrzejszczak commented:
Done in - https://github.com/spring-cloud/spring-cloud-sleuth/commit/19249cdc23a655125a58d022525276e1f74aae09
